sendrecv/gst-rust/src/main.rs

```diff
@@ -0,0 +1,768 @@
+    let candidate = values[2].get::<String>().expect("Invalid argument");
+    let message = serde_json::to_string(&JsonMsg::Ice {
+        candidate: candidate,
+        sdp_mline_index: mlineindex,
+    }).unwrap();
+    app_control.send_text_msg(message.to_string());
+}
+
+fn add_video_source(pipeline: &gst::Pipeline, webrtcbin: &gst::Element) -> Result<(), Error> {
+    let videotestsrc = gst::ElementFactory::make("videotestsrc", None).unwrap();
+    videotestsrc.set_property_from_str("pattern", "ball");
+    let videoconvert = gst::ElementFactory::make("videoconvert", None).unwrap();
+    let queue = gst::ElementFactory::make("queue", None).unwrap();
+    let vp8enc = gst::ElementFactory::make("vp8enc", None).unwrap();
+    vp8enc.set_property("deadline", &1i64)?;
+    let rtpvp8pay = gst::ElementFactory::make("rtpvp8pay", None).unwrap();
+    let queue2 = gst::ElementFactory::make("queue", None).unwrap();
+    pipeline.add_many(&[
+        &videotestsrc,
+        &videoconvert,
+        &queue,
+        &vp8enc,
+        &rtpvp8pay,
+        &queue2,
+    ])?;
+    gst::Element::link_many(&[
+        &videotestsrc,
+        &videoconvert,
+        &queue,
+        &vp8enc,
+        &rtpvp8pay,
+        &queue2,
+    ])?;
+    queue2.link_filtered(webrtcbin, &*RTP_CAPS_VP8)?;
+    Ok(())
+}
+
+fn add_audio_source(pipeline: &gst::Pipeline, webrtcbin: &gst::Element) -> Result<(), Error> {
+    let audiotestsrc = gst::ElementFactory::make("audiotestsrc", None).unwrap();
+    audiotestsrc.set_property_from_str("wave", "red-noise");
+    let queue = gst::ElementFactory::make("queue", None).unwrap();
+    let audioconvert = gst::ElementFactory::make("audioconvert", None).unwrap();
+    let audioresample = gst::ElementFactory::make("audioresample", None).unwrap();
+    let queue2 = gst::ElementFactory::make("queue", None).unwrap();
+    let opusenc = gst::ElementFactory::make("opusenc", None).unwrap();
+    let rtpopuspay = gst::ElementFactory::make("rtpopuspay", None).unwrap();
+    let queue3 = gst::ElementFactory::make("queue", None).unwrap();
+    pipeline.add_many(&[
+        &audiotestsrc,
+        &queue,
+        &audioconvert,
+        &audioresample,
+        &queue2,
+        &opusenc,
+        &rtpopuspay,
+        &queue3,
+    ])?;
+    gst::Element::link_many(&[
+        &audiotestsrc,
+        &queue,
+        &audioconvert,
+        &audioresample,
+        &queue2,
+        &opusenc,
+        &rtpopuspay,
+        &queue3,
+    ])?;
+    queue3.link_filtered(webrtcbin, &*RTP_CAPS_OPUS)?;
+    Ok(())
+}
+
+impl AppControl {
+    fn app_state_eq(&self, state: AppState, error_msg: &'static str) -> bool {
+        if { self.0.lock().unwrap().app_state != state } {
+            self.send_bus_error(String::from(error_msg));
+            false
+        } else {
+            true
+        }
+    }
+
+    fn app_state_lt(&self, state: AppState, error_msg: &'static str) -> bool {
```

Comment:
Maybe this could mirror the `Ord` trait a bit more and be called `app_state_cmp` and return a `std::cmp::Ordering`. But details. Just `lt` seems a bit short :)